src/IO/S3/Client.h

```diff
@@ -0,0 +1,309 @@
+#pragma once
+
+#include "config.h"
+
+#if USE_AWS_S3
+
+#include <Common/logger_useful.h>
+#include <Common/assert_cast.h>
+#include <base/scope_guard.h>
+
+#include <IO/S3/URI.h>
+#include <IO/S3/Requests.h>
+
+#include <aws/core/client/DefaultRetryStrategy.h>
+#include <aws/s3/S3Client.h>
+#include <aws/s3/S3ServiceClientModel.h>
+#include <aws/core/client/AWSErrorMarshaller.h>
+#include <aws/core/client/RetryStrategy.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int TOO_MANY_REDIRECTS;
+}
+
+namespace S3
+{
+
+namespace Model = Aws::S3::Model;
+
+struct ClientCache
+{
+    ClientCache() = default;
+
+    ClientCache(const ClientCache & other)
+        : region_for_bucket_cache(other.region_for_bucket_cache)
+        , uri_for_bucket_cache(other.uri_for_bucket_cache)
```

Comment:
Do we need to copy this cache?

Reply:
No, but I didn't see a reason not to do it. We copy the client for globbed URLs and if it's copied than the endpoint is for sure the same so why not reuse the cache.